File: javascript/26.remove-duplicates-from-sorted-array.js
/*
 * @lc app=leetcode id=26 lang=javascript
 *
 * [26] Remove Duplicates from Sorted Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function (nums) {
  // Solution 1:
  // let i = 0,
  //   k = 0,
  //   temp;
  // while (i < nums.length) {
  //   if (temp !== nums[i]) {
  //     temp = nums[i];
  //     i++;
  //     k++;
  //   } else {
  //     nums.splice(i, 1);
  //   }
  // }
  // return k;

  // Solution 2: slow and fast pointers
  let slow = 0,
    fast = 1;
  while (fast < nums.length) {
    if (nums[fast] !== nums[slow]) {
      nums[++slow] = nums[fast];
    }
    fast++;
  }
  return slow + 1;
};
// @lc code=end

File: rust/26.remove-duplicates-from-sorted-array.rs
/*
 * @lc app=leetcode id=26 lang=rust
 *
 * [26] Remove Duplicates from Sorted Array
 */

// @lc code=start
impl Solution {
    pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
        // Solution 1:
        // nums.dedup();
        // return nums.len() as i32;

        // Solution 2: two pointers
        let (mut s, mut f): (usize, usize) = (0, 1);
        while f < nums.len() {
            if nums[s] != nums[f] {
                s += 1;
                nums.swap(s, f);
            }
            f += 1;
        }
        return (s + 1) as i32;
    }
}
// @lc code=end
